### PR TITLE
Team restore cli command

### PR DESCRIPTION
**Pull request by sonasingh46**
#### Summary
Adds cli for team restore and team archive.

#### Ticket Link
https://github.com/mattermost/mattermost-server/issues/9508
**Note to reviewers:**
Archive CLI command was necessary to add beforehand adding Restore CLI.
Was not very sure about the pattern, raised a WIP PR to get more idea on this and get review comments so as to improve.
Archiving a team command is not throwing any error but not able to figure out what should happen after a team gets archvied.
After running the archive command I can see team in list command intact.
e.g. if team is `myNewTeam`
After archiving and then listing I expected a following output:
`myNewTeam (archived)` but getting `myNewTeam`
Had to add some code in `storetest/mock/TeamStore.go` else was failing(WHY?)

### CODE REVIEW

File: cmd/mattermost/commands/team.go [hunk 1]
```diff
@@ -285,6 +302,54 @@ func searchTeamCmdF(command *cobra.Command, args []string) error {
...
+		return errors.New("Enter at least one team.")
+	}
+
+	teams := getTeamsFromTeamArgs(a, args)
+	for i, team := range teams {
+		if team == nil {
+			CommandPrintErrorln("Unable to find team '" + args[i] + "'")
+			continue
+		}
+		if result := <-a.Srv.Store.Team().SetDeleteAt(team.Id, 0, model.GetMillis()); result.Err != nil {
```
Comment (jespino) [inline cmd/mattermost/commands/team.go]: Normally we use the "App layer" (`app/team.go` in this case) for implement the commands, we don't call directly the store.

File: cmd/mattermost/commands/team.go [hunk 2]
```diff
@@ -285,6 +302,54 @@ func searchTeamCmdF(command *cobra.Command, args []string) error {
...
+	if len(args) < 1 {
+		return errors.New("Enter at least one team to archive.")
+	}
+	teams := getTeamsFromTeamArgs(a, args)
+	for i, team := range teams {
+		if team == nil {
+			CommandPrintErrorln("Unable to find team '" + args[i] + "'")
+			continue
+		}
+		if result := <-a.Srv.Store.Team().Delete(team.Id, model.GetMillis()); result.Err != nil {
```
Comment (jespino) [inline cmd/mattermost/commands/team.go]: We already have implemented the delete in `app/team.go` as `SoftDeleteTeam`. I would use that. And implement a `RestoreTeam` function in `app/team.go` to do the inverse process.

**Comment by jespino** [comment]
Another important thing, this PR must be for the "restore" part only, there os other command for "archive", which is the "soft-delete" process.

**Comment by jespino** [comment]
When you list the teams, there are "all" listed, probably makes sense to include some kind of "(Archived)" flag on them.

**Comment by sonasingh46** [comment]
@jespino 
Sure... I will raise the PR for restore part only.

**Comment by sonasingh46** [comment]
@jespino 
How can I differentiate b/w an archived and non-archived team?
What are the things that cannot be done to a archived team?

**Comment by jespino** [comment]
An archived team have the `DeleteAt` field != 0. For now, an archived teams, from the users perspective (in the application) is almost like you have deleted it. You musn't have access to the team, but for now is something not exposed in the interface, so maybe we have some un-implemented behavior there.

**Comment by sonasingh46** [comment]
@jespino 
Can you look at the PR.

File: app/team.go [hunk 3]
```diff
@@ -1014,6 +1014,19 @@ func (a *App) SoftDeleteTeam(teamId string) *model.AppError {
 	return nil
 }
 
+func (a *App) RestoreTeam(teamId string) *model.AppError {
+	team, err := a.GetTeam(teamId)
+	if err != nil {
+		return err
+	}
+	result := <-a.Srv.Store.Team().SetDeleteAt(team.Id, 0, model.GetMillis())
```
Comment (jespino) [inline app/team.go]: I think here we must follow the same approach of the previous method, set the deleteat to 0 and use the update store method. This way we can remove the store modifications.

**Comment by jespino** [comment]
A comment, but in general looks good to me

File: app/team.go (hunk 3 above)
Comment (sonasingh46) [reply app/team.go]: Will update the PR @jespino

**Comment by sonasingh46** [comment]
@jespino 
Have addressed the review comment.
PTAL.

File: store/sqlstore/team_store.go [hunk 4]
```diff
@@ -980,3 +980,41 @@ func (s SqlTeamStore) GetTeamMembersForExport(userId string) store.StoreChannel
 		result.Data = members
 	})
 }
+
+// Delete records the given deleted timestamp to the channel in question.
```
Comment (jespino) [inline store/sqlstore/team_store.go]: All the store changes can be removed

**Comment by sonasingh46** [comment]
@grundleborg 
Sorry for delay. Was occupied a bit.
Will do necessary changes asap.

**Comment by jasonblais** [comment]
Hey @sonasingh46 - many thanks for your contribution here! This CLI adds a lot of value for admins who would like to archive unused teams on their system.

Looks like we're really close to merging this! If you have any questions on the above feedback, let us know. Happy to help :)

**Comment by sonasingh46** [comment]
@jasonblais 
Sorry.. Was not keeping well so the delay.

**Comment by sonasingh46** [comment]
@jasonblais 
See if the PR

**Comment by sonasingh46** [comment]
@jasonblais 
See if the PR need anything more.

**Comment by grundleborg** [comment]
The build is failing due to the following errors:

```
cmd/mattermost/commands/team.go:81:1: syntax error: unexpected var, expecting expression
cmd/mattermost/commands/team.go:86:31: syntax error: unexpected ) after top level declaration
```

**Comment by jespino** [comment]
@sonasingh46 Yes, would be great if you add unit tests to the command, you can see examples of how to do these in the various *_test.go files in the cmd/mattermost/commands/ directory.

**Comment by grundleborg** [comment]
Hey @sonasingh46 , just wondering if you are still continuing with this PR? I think we are just waiting for a unit test for the new command, and then it should be good to merge.

**Comment by hanzei** [comment]
Hey @sonasingh46,

I appreciate the time you put into this PR. :+1: We just need a unit test for this and we can merge. Do you need help with creating one? Feel free to PM me on the [Mattermost Community Server](https://community.mattermost.com) if you have questions.

If you are not able to finish this PR, I would be glad to finish it for you, if you don't object.

**Comment by sonasingh46** [comment]
@hanzei  
Apologies folks! Got busy with stuff. Thank you for dealing with me. 
Now I have pushed the test code -- it is failing for some reason.
If you can help me fix this.
Also, I have no objection in finishing the PR if this is blocking or prolonged.

**Comment by hanzei** [comment]
@sonasingh46 Thanks for the reply and for adding the test. I went ahead and fixed the build for you. It failed because of some changes we did on `master`.

**Comment by hanzei** [comment]
@wiersgallak Thanks for the review.

@jespino @grundleborg Can you review when you got the time? I would like to get this into 5.8. If you currently don't have the time, just request someone else to review.

File: cmd/mattermost/commands/team.go [hunk 5]
```diff
@@ -294,6 +303,30 @@ func searchTeamCmdF(command *cobra.Command, args []string) error {
...
 }
 
+// Restores archived teams by name
+func restoreTeamsCmdF(command *cobra.Command, args []string) error {
+	a, err := InitDBCommandContextCobra(command)
+	if err != nil {
+		return err
+	}
+	defer a.Shutdown()
+	if len(args) < 1 {
```
Comment (jespino) [inline cmd/mattermost/commands/team.go]: Not a stopper, but we can use `cobra.MinumunNArgs(1)` in the command definition and remove this check.

File: store/storetest/mocks/TeamStore.go [hunk 6]
```diff
@@ -4,9 +4,9 @@
```
Comment (jespino) [inline store/storetest/mocks/TeamStore.go]: Looks like this whole file was accidentally included from a previous iteration.

Comment (hanzei) [reply store/storetest/mocks/TeamStore.go]: Fixed

**Comment by hanzei** [comment]
Thanks for the contribution @sonasingh46! :+1:

**Pull request merged**